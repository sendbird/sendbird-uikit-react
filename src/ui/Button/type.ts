export enum Type {
  PRIMARY,
  SECONDARY,
  DANGER,
  DISABLED,
}

export enum Size {
  BIG,
  SMALL,
}
