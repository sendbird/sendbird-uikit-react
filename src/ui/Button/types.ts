export enum ButtonTypes {
  PRIMARY = 'PRIMARY',
  SECONDARY = 'SECONDARY',
  DANGER = 'DANGER',
  DISABLED = 'DISABLED',
  WARNING = 'WARNING',
}
export enum ButtonSizes {
  BIG = 'BIG',
  SMALL = 'SMALL',
}
