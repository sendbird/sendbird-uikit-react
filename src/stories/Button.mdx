import { Meta, Canvas, Controls, ArgsTable } from '@storybook/addon-docs';
import Button from '../ui/Button';
import * as ButtonStories from './Button.stories';

<Meta title="UI/Button" component={Button} of={ButtonStories} />

# Button

The `Button` component is a versatile element for triggering actions. It supports various sizes, types, disabled states, and can include custom labels. This component is fundamental for creating interactive and accessible user interfaces.

<Canvas of={ButtonStories.Primary} />
<Controls />

### Importing Button

To use the Button component in your project, add the following import statement:

```tsx
import Button from '@sendbird/uikit-react/ui/Button';
```

### Examples
Below are examples showcasing the different configurations of the Button component:

#### Button Types
Shows the Button with different types, indicating various levels of importance or actions.
<Canvas of={ButtonStories.Secondary} />
<Canvas of={ButtonStories.Danger} />

#### Button Sizes
Illustrates the Button in different sizes for fitting into various UI spaces.
<Canvas of={ButtonStories.Big} />
<Canvas of={ButtonStories.Small} />

#### Disabled Button
Demonstrates the Button in a disabled state, showing how it can be used to indicate unavailable actions.
<Canvas of={ButtonStories.Disabled} />
