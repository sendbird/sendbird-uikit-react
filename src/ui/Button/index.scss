@import '../../styles/variables';

.sendbird-button {
  border-radius: 4px;
  box-shadow: none;
  cursor: pointer;
  display: inline-block;
  font-family: var(--sendbird-font-family-default);
  padding: 0;

  &:hover {
    box-shadow: none;
  }

  &:focus {
    outline: none;
  }

  &:active {
    box-shadow: none;
  }

  &.sendbird-button__disabled {
    cursor: not-allowed;
    @include themed() {
      background-color: t(bg-2);
      border: none;
    }

    &:hover {
      @include themed() {
        background-color: t(bg-2);
      }
    }
  }
}

.sendbird-button--big {
  height: 40px;

  .sendbird-button__text {
    margin: 10px 16px;
  }
}

.sendbird-button--small {
  height: 32px;

  .sendbird-button__text {
    margin: 6px 16px;
  }
}

.sendbird-button--primary {
  @include themed() {
    background-color: t(primary-3);
    border: 1px solid t(primary-3);
  }

  .sendbird-button__text {
    @include themed() {
      color: t(on-content-1);
    }
  }

  &:hover {
    @include themed() {
      background-color: t(primary-4);
      border-color: t(primary-4);
    }
  }

  &:focus {
    @include themed() {
      border: 1px solid var(--sendbird-button-border-focus);
      box-shadow: 0 0 0 2px t(primary-3);
    }
  }

  &:active {
    @include themed() {
      background-color: t(primary-5);
    }
  }

  &:focus &:hover {
    @include themed() {
      box-shadow: 0 0 0 2px t(primary-4);
    }
  }

  &:focus &:active {
    @include themed() {
      box-shadow: 0 0 0 2px t(primary-5);
    }
  }
}

.sendbird-button--secondary {
  background-color: transparent;
  @include themed() {
    border: 1px solid t(on-bg-3);
  }

  .sendbird-button__text {
    @include themed() {
      color: t(on-bg-1);
    }
  }

  &:hover {
    background-color: transparent;
    @include themed() {
      border: 1px solid t(primary-3);
    }

    .sendbird-button__text {
      @include themed() {
        color: t(primary-3);
      }
    }
  }

  &.sendbird-button__disabled {
    border: none;
    cursor: not-allowed;
    @include themed() {
      background-color: t(bg-2);
    }

    .sendbird-button__text {
      @include themed() {
        color: t(on-bg-2);
      }
    }

    &:hover {
      border: none;
      @include themed() {
        background-color: t(bg-2);
      }

      .sendbird-button__text {
        @include themed() {
          color: t(on-bg-2);
        }
      }
    }
  }

  &:hover &:focus {
    .sendbird-button__text {
      @include themed() {
        color: t(primary-3);
      }
    }
  }

  &:focus {
    background-color: transparent;
    border: 1px solid transparent;
    @include themed() {
      box-shadow: t(primary-3) 0 0 0 2px;
    }

    .sendbird-button__text {
      @include themed() {
        color: t(on-bg-1);
      }
    }
  }


  &:active {
    box-shadow: none;
    @include themed() {
      background-color: t(bg-1-4);
      border: 1px solid t(primary-3);
    }

    .sendbird-button__text {
      @include themed() {
        color: t(primary-3);
      }
    }
  }
}

.sendbird-button--warning {
  @include themed() {
    background-color: transparent;
    border: 1px solid t(error--3-2);
  }

  .sendbird-button__text {
    @include themed() {
      color: t(on-content-1);
    }
  }

  &:hover {
    @include themed() {
      border-color: t(error-4);
    }
  }

  &:focus {
    @include themed() {
      border-color: var(--sendbird-button-border-focus);
      box-shadow: 0 0 0 2px t(error-3);
    }
  }

  &:hover &:focus {
    @include themed() {
      box-shadow: 0 0 0 2px t(error-4);
    }
  }

  &:active &:focus {
    @include themed() {
      box-shadow: 0 0 0 2px t(error-5);
    }
  }
}

.sendbird-button--danger {
  @include themed() {
    background-color: t(error--3-2);
    border: 1px solid t(error--3-2);
  }

  .sendbird-button__text {
    @include themed() {
      color: t(on-content-1);
    }
  }

  &:hover {
    @include themed() {
      background-color: t(error-4);
      border-color: t(error-4);
    }
  }

  &:focus {
    @include themed() {
      border-color: var(--sendbird-button-border-focus);
      box-shadow: 0 0 0 2px t(error-3);
    }
  }

  &:active {
    box-shadow: none;
    @include themed() {
      background-color: t(error-5);
    }
  }

  &:hover &:focus {
    @include themed() {
      box-shadow: 0 0 0 2px t(error-4);
    }
  }

  &:active &:focus {
    @include themed() {
      box-shadow: 0 0 0 2px t(error-5);
    }
  }
}

.sendbird-button--disabled {
  border: 1px solid transparent;
  cursor: not-allowed;
  @include themed() {
    background-color: t(bg-2);
    color: t(on-content-1);
  }
}

