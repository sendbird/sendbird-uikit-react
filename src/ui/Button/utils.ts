import { ButtonTypes, ButtonSizes } from './types';

export function changeTypeToClassName(type: ButtonTypes): string {
  switch (type) {
    case ButtonTypes.PRIMARY: return 'sendbird-button--primary';
    case ButtonTypes.SECONDARY: return 'sendbird-button--secondary';
    case ButtonTypes.DANGER: return 'sendbird-button--danger';
    case ButtonTypes.DISABLED: return 'sendbird-button--disabled';
    case ButtonTypes.WARNING: return 'sendbird-button--warning';

    default: return null;
  }
}

export function changeSizeToClassName(size: ButtonSizes): string {
  switch (size) {
    case ButtonSizes.BIG: return 'sendbird-button--big';
    case ButtonSizes.SMALL: return 'sendbird-button--small';
    default: return null;
  }
}
