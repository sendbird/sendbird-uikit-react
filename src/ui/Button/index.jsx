import React from 'react';
import PropTypes from 'prop-types';

import './index.scss';
import Label, { LabelTypography, LabelColors } from '../Label';
import { Type, Size } from './type';
import { changeTypeToClassName, changeSizeToClassName } from './utils';

export default function Button({
  className,
  type,
  size,
  children,
  disabled,
  onClick,
}) {
  const injectingClassNames = [
    ...((Array.isArray(className)) ? className : [className]),
    'sendbird-button',
    (disabled ? 'sendbird-button__disabled' : ''),
    changeTypeToClassName(type),
    changeSizeToClassName(size),
  ].join(' ');

  return (
    <button
      className={injectingClassNames}
      type="button"
      onClick={onClick}
      disabled={disabled}
    >
      <Label
        className="sendbird-button__text"
        type={LabelTypography.BUTTON_1}
        color={LabelColors.ONCONTENT_1}
      >
        {children}
      </Label>
    </button>
  );
}

export const ButtonTypes = Type;
export const ButtonSizes = Size;

Button.propTypes = {
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),
  type: PropTypes.oneOf([
    'PRIMARY',
    'SECONDARY',
    'DANGER',
    'DISABLED',
  ]),
  size: PropTypes.oneOf(['BIG', 'SMALL']),
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]),
  disabled: PropTypes.bool,
  onClick: PropTypes.func,
};

Button.defaultProps = {
  className: '',
  type: Type.PRIMARY,
  size: Size.BIG,
  children: 'Button',
  disabled: false,
  onClick: () => { },
};
