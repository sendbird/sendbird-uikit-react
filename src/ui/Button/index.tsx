import React, { ReactElement } from 'react';

import './index.scss';
import Label, { LabelTypography, LabelColors } from '../Label';
import { Type, Size } from './type';
import { changeTypeToClassName, changeSizeToClassName } from './utils';

export interface ButtonProps {
  className?: string | Array<string>;
  type: Type;
  size: Size;
  children: string | ReactElement;
  disabled?: boolean;
  onClick?: () => void;
}

export default function Button({
  className,
  type = Type.PRIMARY,
  size = Size.BIG,
  children = 'Button',
  disabled = false,
  onClick = () => {/* noop */},
}: ButtonProps): ReactElement {
  const injectingClassNames = [
    ...((Array.isArray(className)) ? className : [className]),
    'sendbird-button',
    (disabled ? 'sendbird-button__disabled' : ''),
    changeTypeToClassName(type),
    changeSizeToClassName(size),
  ].join(' ');

  return (
    <button
      className={injectingClassNames}
      type="button"
      onClick={onClick}
      disabled={disabled}
    >
      <Label
        className="sendbird-button__text"
        type={LabelTypography.BUTTON_1}
        color={LabelColors.ONCONTENT_1}
      >
        {children}
      </Label>
    </button>
  );
}

export const ButtonTypes = Type;
export const ButtonSizes = Size;
