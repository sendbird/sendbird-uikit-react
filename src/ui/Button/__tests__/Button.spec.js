import React from 'react';
import { render, screen } from '@testing-library/react';

import Button, { ButtonSizes } from "../index";

describe('ui/Button', () => {
  it('should do a snapshot test of the default Button DOM', function () {
    const { asFragment } = render(<Button />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should do a snapshot test of the small Button DOM', function () {
    const { asFragment } = render(<Button size={ButtonSizes.SMALL} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should contain className', function () {
    const text = 'example-text';
    render(<Button className={text} />);
    expect(screen.getByRole('button').className).toContain(text);
  });
});
