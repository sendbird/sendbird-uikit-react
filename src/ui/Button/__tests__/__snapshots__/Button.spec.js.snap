// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should do a snapshot test of the default Button DOM 1`] = `
<button
  className=" sendbird-button  sendbird-button--primary sendbird-button--big"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <span
    className="sendbird-button__text sendbird-label sendbird-label--button-1 sendbird-label--color-oncontent-1"
  >
    Button
  </span>
</button>
`;

exports[`Button should do a snapshot test of the small Button DOM 1`] = `
<button
  className=" sendbird-button  sendbird-button--primary sendbird-button--small"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <span
    className="sendbird-button__text sendbird-label sendbird-label--button-1 sendbird-label--color-oncontent-1"
  >
    Button
  </span>
</button>
`;
